### PR TITLE
Nikita/database

### PR DESCRIPTION
**Pull request by NikitaShymberg**
TODO: tests
TODO: small renames/refactors

(Finally) a PR! This implements the queries to be used by the server to access the database. I am expecting a lot of comments w.r.t. both code quality and compatibility with @kulvirs code so fire away. I'm pretty sure I've got all the necessary functions down, but let me know if I've missed any.

I also seem to have fixed the issue that I've been complaining about - it seems to have been more of an sqlite problem than a library problem (although the library is not without blame).

### CODE REVIEW

**Review by okjuan — CHANGES_REQUESTED** [review]
Good job dude, this is looking good! You might be sorta blocked by #2 since you need to use the `Database` interface and `Changes` object declared in there. Specifically, the `Changes` object is not yet finalized (see relevant discussion [here](https://github.com/Dabblr/Concurrent-Document-Editor/pull/2#discussion_r232477512)). I'd suggest you rebase onto `kulvir/server` for the time being, until that one is merged to `master`. (I can help with this, it can be done in just a couple of commands 😃)

Some initial suggestions:
- Incorporate Kulvir's changes:
  - Move stuff to the `database/` package
  - Use `Change` object (defined by Kulvir) by importing it
  - Implement the `Database` interface by implementing the functions verbatim (first char of function name should be uppercase and types should match)
    - Might have to create a new type (could be as simple as a struct with a string field for the db   path/name) in order to implement the Database interface; then, you could add pointer receivers to  your functions

- Testing
  - Move the stuff in main function to a _test.go file and add more tests
  - Maybe install a Golang linter to get suggestions about code style (e.g. function comments for exported functions)

Sorry, I know this is a lot! It might be easier to chat in person to clarify some stuff here. Let me know what you think!

**Comment by kulvirs** [comment]
The function `createEmptyFile()` needs to be changed to return a file ID, revision number, and an error in order to match up with what is in master right now and resolve merge conflicts.

File: database/databaseinterface.go
Comment (kulvirs) [inline database/databaseinterface.go]: The function `createEmptyFile()` needs to be changed to return a file ID, revision number, and an error in order to match up with what is in master right now and resolve merge conflicts

**Review by kulvirs — CHANGES_REQUESTED** [review]
The function `createEmptyFile()` needs to be changed to return a file ID, revision number, and an error in order to match up with what is in master right now and resolve merge conflicts

**Comment by kulvirs** [comment]
It looks like if we merged this right now, since it doesn't have the most recent changes from master, it might write over some of the later changes I made in #2 to `mockDB.go` , `applyUpdates.go`, and `server.go`.

**Comment by NikitaShymberg** [comment]
I see what went wrong, I didn't realise that `git fetch` didn't fetch all branches, so I needed to check out master and pull first. This should be much better now!

File: database/database.go
Comment (okjuan) [inline database/database.go:193]: At this point, do we know for sure that `revs` is not `nil`?

File: database/database.go
Comment (okjuan) [inline database/database.go]: It looks like this `if` statement will always evaluate to true.

File: database/database_test.go
Comment (okjuan) [inline database/database_test.go:70]: `f` might be `nil` if we `err` is `nil`

**Review by kulvirs — APPROVED** [review]

**Comment by kulvirs** [comment]
Looks good to merge to me!

File: database/database_test.go
Comment (okjuan) [inline database/database_test.go]: What's the `Real` suffix for? 😄

File: database/createDb.sql
Comment (okjuan) [inline database/createDb.sql]: Line breaks would make this a bit easier to read :-)
e.g.
```suggestion
CREATE TABLE changes (
  file integer,
  rev_number integer,
  position number,
  character CHARACTER,
  FOREIGN KEY(file) references revisions(file),
  FOREIGN KEY(rev_number) references revisions(number),
  PRIMARY KEY(file, rev_number, position)
);
```

File: database/database.go
Comment (okjuan) [inline database/database.go]: Line breaks here too would make things a bit cleaner.
Something like:
```suggestion
	revs, err := conn.Query(`
	 SELECT DISTINCT changes.file, rev_number, position, character
	 FROM revisions JOIN changes
	 WHERE changes.file = ? AND number > ?`, id, revisionNumber)
```

If you do this, you might have to play around a bit before the compiler is happy. Note that the [back-tick allows multi-line strings](https://stackoverflow.com/questions/7933460/how-do-you-write-multiline-strings-in-go).

File: database/database.go
Comment (okjuan) [inline database/database.go:95]: A comment about the auto-generated `ROW_ID` column would be great

File: database/database.go
Comment (okjuan) [inline database/database.go:59]: Same here, a comment about the auto-generated `ROW_ID` column would be great

File: database/createDb.sql
Comment (okjuan) [inline database/createDb.sql]: Comments about `ROW_ID` and its relevance to the two `id` columns would be great!

File: database/database.go
Comment (okjuan) [inline database/database.go]: Could you update the comment regarding the return type so that it explains what the second `int` being returned is?

File: database/database.go
Comment (okjuan) [inline database/database.go:34]: Since it's repeated in various places, maybe the logic for opening a connection (including checking for error) could be put into a helper function (e.g. `getConnection()`)? Totally up to you. I think you'd still have to invoke `conn.Close()` after calling that new helper function anyway.

File: database/database.go
Comment (okjuan) [inline database/database.go:199]: I see that deletions are identified by an empty `character` column. This seems a bit risky because we are currently relying on the caller of `InsertChanges` to do so. So I'd suggest we do one of two things:
- Add a `type varchar(10)` column to the `changes` table so that we can check what the change is that way, or
- Make sure that deletions always have an empty `character` in the database (by adding logic in `InsertChanges` to do so)

Thoughts?

File: database/createDb.sql
Comment (okjuan) [inline database/createDb.sql]: We are allowing multiple files to have the same name, which I think makes sense. However, we should probably enforce uniqueness on the tuple `(owner, filename)` so that a single user cannot have multiple files with the same name. Otherwise, we'd need some other way to disambiguate them.

File: database/database_test.go
Comment (okjuan) [inline database/database_test.go]: It's great that we're testing this 👍however, I think we'd normally want our tests to run independently of each other, so that no one individual test relies on another test in order to pass. 

Of course, we have various test scenarios where we want to put the database in a specific state so that we can test the behaviour of the `database.go` API in that context. Typically, this is done in setup and tear-down methods, which are called immediately before and after each test function. Golang supports this kind of behaviour in a simpler way: check it out [here](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks). Basically, you just have a test function that calls other test functions but does so using `t.Run()`.

File: database/database_test.go
Comment (okjuan) [inline database/database_test.go:14]: In the spirit of keeping making tests independent of each other (see my other long comment in this file lol), I'd suggest avoiding global variables and declaring variables like these anew inside each test function as you need them.

Although, this is more about style and less of a big deal than having tests tests not depend on each other's behaviour.

File: sql.go
Comment (okjuan) [inline sql.go]: Do we still need this file?

File: database/databaseinterface.go
Comment (okjuan) [inline database/databaseinterface.go:10]: Could you revert the comment to mention the revision number? 😄 (unless we don't wanna return it?)

**Review by okjuan — COMMENTED** [review]
Great job dude!! 👍 💣 💪 🔥 

I left a bunch of suggestions, most of which are entirely optional and a few of which we should turn into GitHub issues if they're addressed in this PR. I'll approve this bad boy in a bit, I'll just test it and maybe run it locally as well.

Couple last notes:
* I don't think we wanna track the actual database with git btw. I think we'll eventually write a deployment script that involves creating the database and starting the server.
* I think we should update the path from `github.com/jcgallegdup` to `github.com/Dabblr` (like Kulvir did in her PR)
* We currently aren't handling the return value from the `.exec()` function. We might wanna do that since it [returns an error](https://godoc.org/github.com/mxk/go-sqlite/sqlite3#Stmt.Exec)
* The column `username` in `users` table should probably be unique

File: database/database_test.go
Comment (kulvirs) [inline database/database_test.go]: Maybe you should add a test that verifies CreateEmptyFile returns a revision id of 1?

**Review by kulvirs — CHANGES_REQUESTED** [review]
I'm getting the following errors (2 tests failing) when I try to run your unit tests for the database. Not sure if this is because I'm using sqlite on Windows which is causing some difference with my local database or what, but this is probably something worth looking into. 
```
--- FAIL: TestGetFileContentReturnsMatchingContentReal (0.02s)
    database_test.go:59: GetFileContent threw and error sqlite3: no such column: name [1]
    database_test.go:62: Incorrect file content, received content:
c: Ins: pos=0 val=a
c: Ins: pos=1 val=b
F0 Ins: pos=0 val=a
F1 Ins: pos=1 val=b
--- FAIL: TestGetChangesSinceRevisionReturnsEmptyArrayIfNoChangesReal (0.02s)
    database_test.go:107: GetChangesSinceRevision was supposed to return an empty array. Instead we got [Ins: pos=0 val=a Ins: pos=1 val=b]
c: Ins: pos=0 val=a
c: Ins: pos=1 val=b
2018/12/06 22:05:34 New file content: newContent
FAIL
exit status 1
FAIL    github.com/Dabblr/Concurrent-Document-Editor/database   2.552s
```

**Comment by kulvirs** [comment]
I updated `mockDB.go`, `mockDB_test.go`, `server.go` and `update.go` to make sure the mock database follows the new database interface and the server addresses potential errors being returned from the database functions.   
Aside from two of the database tests written by @NikitaShymberg that I mentioned are failing in the comment above, all other tests now pass and the server runs fine with the mock database.

File: database/database.go
Comment (NikitaShymberg) [reply database/database.go:193]: Good catch!

Comment (NikitaShymberg) [reply database/database.go:193]: **Fixed**. This was actually kind of interesting as `revs == nil` if the file isn't found _and_ if the given revision is the latest revision, one of these was intended and the other not. So I added an extra check to see if the file exists.

File: database/database.go
Comment (NikitaShymberg) [reply database/database.go]: **Fixed** I think I started future proofing against something but it turned out to be unnecessary/

File: database/database_test.go
Comment (NikitaShymberg) [reply database/database_test.go:70]: I don't think so, `GetFileContent` always returns a file object and it looks like in Go that will never be `nil`. All of its fields will be empty though.

File: database/database_test.go
Comment (NikitaShymberg) [reply database/database_test.go]: Uh yeah this one's a little awkward. Go doesn't allow duplicate functions with the same name to be exported, since a lot of these tests are replicated in the `mockDB_test.go` file there were naming conflicts. Do you have any suggestions for better names? Or maybe another solution?

File: database/createDb.sql
Comment (NikitaShymberg) [reply database/createDb.sql]: **Fixed.** Good point.

File: database/database.go
Comment (NikitaShymberg) [reply database/database.go]: **Fixed.** Good point, I did leave some of the super short queries on one line.

File: database/database.go
Comment (NikitaShymberg) [reply database/database.go]: **Fixed.** Oh yup, I think @kulvirs asked for it to also return the latest revision which will always be 0.

File: database/database.go
Comment (NikitaShymberg) [reply database/database.go:34]: I definitely thought about this, but I couldn't think of a clean way of doing it. As you said the `defer` statement should always be in this outer function so that line is going to have to stay. The more problematic thing is that in the event of an error each function returns a different thing (e.g `CreateUser` returns -1 as the `userID`, whereas `GetFileContent` returns a dummy file).

File: database/createDb.sql
Comment (NikitaShymberg) [reply database/createDb.sql]: I'm not sure if that's necessary, since each file already has a unique ID it should be fine as far as the database goes. As far as user experience goes, Google Docs lets you create multiple docs with the same name so ¯\\\_(ツ)\_/¯

File: database/database_test.go
Comment (NikitaShymberg) [reply database/database_test.go]: As we discussed last Sunday, I'll do this in another PR unless we run out of time!

File: sql.go
Comment (NikitaShymberg) [reply sql.go]: **Fixed.** Nope, that came from the merge zzz

File: database/database_test.go
Comment (NikitaShymberg) [reply database/database_test.go:14]: This one I think I disagree with. Imo it makes it cleaner to not have to re-declare constants inside each test case, especially frequently used ones. Source: they encouraged us to create global constants for tests at my last coop.

File: database/databaseinterface.go
Comment (NikitaShymberg) [reply database/databaseinterface.go:10]: **Fixed.**

File: database/database_test.go
Comment (NikitaShymberg) [reply database/database_test.go]: **Fixed.**

File: database/createDb.sql
Comment (NikitaShymberg) [reply database/createDb.sql]: Can you give an example of what you mean for these?

**Comment by NikitaShymberg** [comment]
Tests are being a butt :cry: The last test `TestGetChangesSinceRevisionReturnsChangeArrayReal` seems to fail most of the time if all the database tests are run at once, but running it one more time makes it pass :man_shrugging: . I think I want to fix them in a future PR that will focus entirely on the tests.

File: database/database_test.go
Comment (kulvirs) [reply database/database_test.go]: Oh yeah, that's partially my fault. If you want, you could change all my tests in `mockDB_test.go` to start with `TestMock` and then you could get rid of `Real` in your tests. I think that would make more sense.

File: database/database_test.go
Comment (kulvirs) [reply database/database_test.go:14]: I think having some global constants that are common to all tests is fine as well. The main thing is just that each test should pass/fail independently of whether another test does.

**Comment by kulvirs** [comment]
Hmm I wonder if the reason the database tests fail like that is because Go tries to be efficient and run the tests in parallel? In which case, maybe the problem could be fixed by making each test independent of the others? 
Gotta love concurrency bugs. So much uncertainty, you know. :smirk:

**Comment by NikitaShymberg** [comment]
That's what I was thinking too @kulvirs, after a (quick) google around I wasn't able to find anything to confirm/deny this though so I am uncertain. Given that most of the problems are coming from only the tests, I propose that we merge this PR in so that we can have a productive day tomorrow. After we've each prepared the presentation/report, I will make a PR for tests. I know that this isn't the best idea for production code but given the time constraints (and my newness to unit tests) I think we should prioritise getting a minimum viable product.

File: database/database_test.go
Comment (okjuan) [reply database/database_test.go:14]: I'm ok with that 👍

File: database/createDb.sql
Comment (okjuan) [reply database/createDb.sql]: Maybe something like this:
```suggestion
-- 'id' column is an alias for SQLite's ROW_ID column, which is automatically populated
CREATE TABLE files (id integer PRIMARY KEY, filename STRING NOT NULL, data BLOB, owner integer, FOREIGN KEY(owner) REFERENCES users(id));
```

File: database/database.go
Comment (okjuan) [reply database/database.go:34]: If an error occurs, I think we don't return anything anyway since we call `panic()`; or are you referring to a different error?

File: database/database_test.go
Comment (okjuan) [reply database/database_test.go]: OH right, I see. Naming the mock ones differently might be slightly better, but I'll just leave it up to you

**Review by kulvirs — APPROVED** [review]

**Pull request merged**
